File: dotnet/src/SodiumFRP/CSharp/SodiumFRP/Listener.cs
using System.Collections.Generic;

namespace SodiumFRP
{
    public static class Listener
    {
        public static readonly IListener Empty = ListenerInternal.EmptyImpl;

        public static IListener CreateComposite(IReadOnlyList<IListener> listeners) => ListenerInternal.CreateCompositeImpl(listeners);

        public static IWeakListener CreateWeakComposite(IReadOnlyList<IWeakListener> listeners) =>
            ListenerInternal.CreateWeakCompositeImpl(listeners);

        public static IStrongListener CreateStrongComposite(IReadOnlyList<IStrongListener> listeners) =>
            ListenerInternal.CreateStrongCompositeImpl(listeners);

        public static IListener Append(IListener listener1, IListener listener2) =>
            ListenerInternal.AppendImpl(listener1, listener2);
    }
}

File: dotnet/src/SodiumFRP/FSharp/SodiumFRP.FSharp/Listener.fs
module SodiumFRP.Listener

open System.Runtime.CompilerServices

let unlistenWeak (listener : IListenerWithWeakReference) = listener.Unlisten ()
let unlisten (listener : IListener) = listener.Unlisten ()
let getListenerWithWeakReference (listener : IListener) = listener.GetListenerWithWeakReference ()

[<MethodImpl(MethodImplOptions.NoInlining)>]
let empty = ListenerInternal.EmptyImpl

[<MethodImpl(MethodImplOptions.NoInlining)>]
let fromList listeners = ListenerInternal.CreateCompositeImpl listeners

[<MethodImpl(MethodImplOptions.NoInlining)>]
let fromWeakList listeners = ListenerInternal.CreateWeakCompositeImpl listeners

[<MethodImpl(MethodImplOptions.NoInlining)>]
let fromStrongList listeners = ListenerInternal.CreateStrongCompositeImpl listeners

let fromSeq listeners = List.ofSeq listeners |> fromList
let fromWeakSeq listeners = List.ofSeq listeners |> fromWeakList
let fromStrongSeq listeners = List.ofSeq listeners |> fromStrongList

[<MethodImpl(MethodImplOptions.NoInlining)>]
let append l1 l2 = ListenerInternal.AppendImpl (l1, l2)